File: app/backend/src/services/genAiService.js
import { env } from '../config/env.js';

const promptVersion = 'business-model-rec-v1';
const suggestionPromptVersion = 'business-model-suggestion-v1';

function compactImpact(impact = {}) {
  const o2c = Object.entries(impact.o2cImpactScores || {})
    .map(([area, value]) => ({ area, score: value.score ?? value, severity: value.severity }))
    .sort((a, b) => Number(b.score || 0) - Number(a.score || 0))
    .slice(0, 6);
  return {
    scenarioId: impact.scenarioId,
    complianceRisk: impact.complianceRisk,
    implementationEffort: impact.implementationEffort,
    revenueImpact: impact.revenueImpact,
    delayProbability: impact.delayProbability,
    confidenceScore: impact.confidenceScore,
    integrationImpactScores: impact.integrationImpactScores,
    topO2CAreas: o2c
  };
}

function compactPrediction(prediction = {}) {
  const values = prediction.predictions || prediction;
  return {
    predictions: values,
    confidence: prediction.confidence,
    modelVersion: prediction.modelVersion,
    trainingMetadata: prediction.trainingMetadata,
    topContributingFactors: (prediction.topContributingFactors || []).slice(0, 6)
  };
}

function compactPredictiveAnalysis(analysis = {}) {
  return {
    modelVersion: analysis.modelVersion,
    confidence: analysis.confidence,
    historicalPatternSummary: analysis.historicalPatternSummary,
    topContributingFactors: (analysis.topContributingFactors || []).slice(0, 6),
    predictedOutputs: analysis.predictedOutputs
  };
}

function compactImpactDetails(details = {}) {
  return {
    scenarioOverview: details.scenarioOverview,
    predictiveAi: details.predictiveAi,
    riskMetrics: (details.riskMetrics || []).slice(0, 8),
    areaDetails: (details.areaDetails || []).slice(0, 8).map((area) => ({
      label: area.label,
      score: area.score,
      severity: area.severity,
      reason: area.reason,
      affectedSystems: area.affectedSystems,
      dependencyCount: area.dependencyCount
    }))
  };
}

function buildPrompt({ scenario, impact, prediction, predictiveAnalysis, comparison }) {
  return [
    'You are an enterprise commercial model transformation advisor for SAP Order-to-Cash programs.',
    'Return only valid compact JSON. No markdown.',
    'Generate a concise JSON object with keys executiveSummary, recommendation, rationale, risks, assumptions, businessImpact, implementationConsiderations, whyThisModel.',
    'Base every statement on the supplied scenario, O2C impact analysis, local predictive AI historical analysis, predictive model output, contributing factors and comparison scores. Do not invent external facts.',
    'Use the predictive AI outputs as the quantitative evidence and write the recommendation as business-readable reasoning.',
    `Scenario: ${JSON.stringify(scenario)}`,
    `Impact analysis: ${JSON.stringify(compactImpact(impact))}`,
    `Predictive output: ${JSON.stringify(compactPrediction(prediction))}`,
    `Predictive historical analysis: ${JSON.stringify(compactPredictiveAnalysis(predictiveAnalysis || {}))}`,
    `Comparison: ${JSON.stringify(comparison || [])}`
  ].join('\n');
}

function unavailableRecommendation(details) {
  return {
    success: false,
    status: 'GENAI_UNAVAILABLE',
    source: 'unavailable',
    promptVersion,
    model: env.openSourceAi.model,
    message: 'AI recommendation is temporarily unavailable.',
    customerMessage: 'We could not generate a new AI recommendation right now. Predictive impact scores and historical analysis are still available. Please retry after checking the AI service configuration.',
    technicalDetails: details || 'The configured GenAI endpoint is unavailable or not configured.',
    predictiveOnly: true,
    suggestedScenario: null,
    executiveSummary: '',
    recommendation: '',
    rationale: '',
    risks: [],
    assumptions: [],
    businessImpact: '',
    implementationConsiderations: '',
    whyThisModel: '',
    generatedAt: new Date().toISOString()
  };
}

function parseModelContent(body) {
  return body?.choices?.[0]?.message?.content
    || body?.message?.content
    || body?.response
    || body?.output_text
    || body?.content?.[0]?.text
    || JSON.stringify(body);
}

function parseJsonContent(content) {
  const trimmed = content.trim();
  const fenced = trimmed.match(/^```(?:json)?\s*([\s\S]*?)\s*```$/i);
  const candidate = fenced ? fenced[1] : trimmed;
  try {
    return JSON.parse(candidate);
  } catch {
    return { executiveSummary: content };
  }
}

function buildSuggestionPrompt({ scenario, impact, prediction, impactDetails, comparisonContext }) {
  const values = prediction?.predictions || prediction || {};
  return [
    'You are an enterprise SAP Order-to-Cash commercial model architect.',
    'Suggest a better business model only if the supplied evidence indicates a measurable improvement opportunity.',
    'Return only valid JSON. No markdown, no prose outside JSON.',
    'The JSON must match this schema:',
    JSON.stringify({
      noBetterModelFound: false,
      message: '',
      suggestedScenario: {
        name: '',
        businessModelType: '',
        pricingType: '',
        billingFrequency: '',
        contractTerm: 0,
        fundingModel: '',
        bundleType: '',
        transactionVolume: 0,
        expectedRevenue: 0,
        complianceRegions: [],
        integrationCount: 0,
        processComplexity: 0
      },
      improvementSummary: '',
      aiRationale: '',
      expectedBenefits: {
        revenueImprovement: 0,
        riskReduction: 0,
        effortChange: 0,
        delayReduction: 0
      },
      o2cImpactChanges: [],
      tradeOffs: [],
      assumptions: [],
      confidence: 0
    }),
    'If no better model exists, return {"noBetterModelFound":true,"message":"The current model appears optimal based on the available impact and prediction data.","aiRationale":"","confidence":0}.',
    'Allowed businessModelType values: subscription, consumption-based, tiered, bundled-services, outcome-based, flexible-funding, hybrid.',
    'Allowed pricingType values: flat, tiered, usage, outcome, bundle, hybrid.',
    'Allowed billingFrequency values: monthly, quarterly, annual, milestone, usage-event.',
    'Allowed fundingModel values: capex, opex, mixed, partner-funded, customer-funded.',
    'Allowed bundleType values: single-offer, software-service, hardware-software-service, multi-product, partner-ecosystem.',
    'Use only scenario data, impact analysis, predictive output, O2C affected areas, compliance regions, integration complexity, revenue expectations and comparison criteria.',
    'If comparison context includes previousSuggestedModel entries, avoid repeating the same model, pricing type and billing frequency unless the predictive evidence clearly proves it remains best.',
    'Suggested values must be adapted to the submitted scenario; do not reuse a generic recommendation pattern across unrelated industries, customer segments, regions or O2C risk profiles.',
    'Only return noBetterModelFound when implementationRisk, billingComplexity, operationalEffort, implementationEffort and delayProbability are all moderate or low.',
    `Evidence threshold check: ${JSON.stringify({
      mustSuggestAlternative: evidenceRequiresAlternative({ scenario, impact, prediction }),
      implementationRisk: values.implementationRisk,
      billingComplexity: values.billingComplexity,
      operationalEffort: values.operationalEffort,
      implementationEffort: impact?.implementationEffort,
      delayProbability: values.delayProbability ?? impact?.delayProbability,
      processComplexity: scenario?.processComplexity,
      integrationCount: scenario?.integrationCount
    })}`,
    `Original scenario: ${JSON.stringify(scenario)}`,
    `Impact analysis: ${JSON.stringify(compactImpact(impact))}`,
    `Predictive AI output: ${JSON.stringify(compactPrediction(prediction))}`,
    `Impact details: ${JSON.stringify(compactImpactDetails(impactDetails))}`,
    `Comparison context: ${JSON.stringify(comparisonContext || [])}`
  ].join('\n');
}

function evidenceRequiresAlternative({ scenario = {}, impact = {}, prediction = {} }) {
  const values = prediction.predictions || prediction || {};
  return Number(values.implementationRisk || 0) >= 60
    || Number(values.billingComplexity || 0) >= 70
    || Number(values.operationalEffort || 0) >= 70
    || Number(values.delayProbability ?? impact.delayProbability ?? 0) >= 50
    || Number(impact.implementationEffort || 0) >= 70
    || Number(scenario.processComplexity || 0) >= 7
    || Number(scenario.integrationCount || 0) >= 12;
}

function noBetterResult(parsed) {
  return {
    success: true,
    source: env.openSourceAi.provider,
    promptVersion: suggestionPromptVersion,
    model: env.openSourceAi.model,
    noBetterModelFound: true,
    message: parsed.message || 'The current model appears optimal based on the available impact and prediction data.',
    aiRationale: parsed.aiRationale || '',
    confidence: Number(parsed.confidence || 0),
    generatedAt: new Date().toISOString()
  };
}

function validateSuggestionShape(parsed, requiresAlternative) {
  if (parsed.noBetterModelFound) {
    if (requiresAlternative) throw new Error('Predictive evidence requires an alternative suggestion');
    return noBetterResult(parsed);
  }
  if (!parsed.suggestedScenario || !parsed.expectedBenefits) {
    throw new Error('Missing suggestedScenario or expectedBenefits');
  }
  return null;
}

function aiRequestBody(prompt, maxTokens) {
  const body = {
    model: env.openSourceAi.model,
    messages: [{ role: 'user', content: prompt }],
    stream: false,
    think: false,
    format: 'json',
    options: {
      num_predict: maxTokens,
      temperature: 0.2
    }
  };
  if (!isNativeOllama()) {
    body.max_tokens = maxTokens;
    body.temperature = 0.2;
    body.response_format = { type: 'json_object' };
  }
  return body;
}

function isNativeOllama() {
  return env.openSourceAi.provider === 'ollama' && /localhost:11434|127\.0\.0\.1:11434/.test(env.openSourceAi.baseUrl);
}

function aiEndpointUrl() {
  if (!isNativeOllama()) return env.openSourceAi.baseUrl;
  return env.openSourceAi.baseUrl
    .replace(/\/v1\/chat\/completions\/?$/, '/api/chat')
    .replace(/\/v1\/?$/, '/api/chat');
}

function createAiError(errorCode, message, details, status = 503) {
  const error = new Error(message);
  error.status = status;
  error.errorCode = errorCode;
  error.details = details;
  return error;
}

function validateAiConfig() {
  if (!env.openSourceAi.baseUrl || !env.openSourceAi.model) {
    throw createAiError(
      'GENAI_UNAVAILABLE',
      'AI suggestion service is not configured.',
      'Configure SAP AI Core or an OpenAI-compatible GenAI endpoint before requesting suggestions.'
    );
  }
}

async function callConfiguredAiOnce(prompt, maxTokens = 1200) {
  validateAiConfig();

  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), env.openSourceAi.timeoutMs);
  const headers = { 'Content-Type': 'application/json' };
  if (env.openSourceAi.apiKey) {
    headers.Authorization = `Bearer ${env.openSourceAi.apiKey}`;
  }

  try {
    const response = await fetch(aiEndpointUrl(), {
      method: 'POST',
      signal: controller.signal,
      headers,
      body: JSON.stringify(aiRequestBody(prompt, maxTokens))
    });
    if (!response.ok) {
      throw createAiError(
        'GENAI_PROVIDER_ERROR',
        'AI provider returned an unsuccessful response.',
        `The GenAI service returned HTTP ${response.status}.`,
        502
      );
    }
    const body = await response.json();
    return parseModelContent(body);
  } catch (error) {
    if (error.name === 'AbortError') {
      console.warn('AI request timed out or was aborted', {
        provider: env.openSourceAi.provider,
        timeoutMs: env.openSourceAi.timeoutMs,
        errorName: error.name
      });
      throw createAiError(
        'AI_REQUEST_TIMEOUT',
        'AI suggestion request timed out. Please retry.',
        'The GenAI service did not respond within the configured timeout.',
        504
      );
    }
    if (error.errorCode) throw error;
    console.warn('AI request failed', {
      provider: env.openSourceAi.provider,
      errorName: error.name,
      message: error.message
    });
    throw createAiError(
      'GENAI_NETWORK_ERROR',
      'AI suggestion request failed due to a network or provider connection issue.',
      'Check that the configured GenAI endpoint is reachable and retry.',
      503
    );
  } finally {
    clearTimeout(timeout);
  }
}

async function callConfiguredAi(prompt, maxTokens = 1200, retries = 1) {
  let lastError;
  for (let attempt = 0; attempt <= retries; attempt += 1) {
    try {
      return await callConfiguredAiOnce(prompt, maxTokens);
    } catch (error) {
      lastError = error;
      if (!['AI_REQUEST_TIMEOUT', 'GENAI_NETWORK_ERROR'].includes(error.errorCode) || attempt === retries) {
        throw error;
      }
    }
  }
  throw lastError;
}

export async function generateRecommendation(input) {
  const prompt = buildPrompt(input);
  if (!env.openSourceAi.baseUrl || !env.openSourceAi.model) {
    return unavailableRecommendation('Configure SAP AI Core or an OpenAI-compatible GenAI endpoint before requesting recommendations.');
  }

  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), env.openSourceAi.timeoutMs);
  const headers = { 'Content-Type': 'application/json' };
  if (env.openSourceAi.apiKey) {
    headers.Authorization = `Bearer ${env.openSourceAi.apiKey}`;
  }

  try {
    const response = await fetch(aiEndpointUrl(), {
      method: 'POST',
      signal: controller.signal,
      headers,
      body: JSON.stringify(aiRequestBody(prompt, 220))
    });
    if (!response.ok) throw new Error(`${env.openSourceAi.provider} returned ${response.status}`);
    const body = await response.json();
    const parsed = parseJsonContent(parseModelContent(body));
    return {
      source: env.openSourceAi.provider,
      promptVersion,
      model: env.openSourceAi.model,
      ...parsed,
      generatedAt: new Date().toISOString()
    };
  } catch (error) {
    return unavailableRecommendation(error.message || 'The configured GenAI endpoint could not be reached.');
  } finally {
    clearTimeout(timeout);
  }
}

export async function generateBetterModelSuggestion(input) {
  const prompt = buildSuggestionPrompt(input);
  const requiresAlternative = evidenceRequiresAlternative(input);
  let content;
  try {
    content = await callConfiguredAi(prompt, 320, 1);
  } catch (error) {
    const controlled = new Error(error.message);
    controlled.status = error.status || 503;
    controlled.errorCode = error.errorCode || 'GENAI_UNAVAILABLE';
    controlled.details = error.details;
    throw controlled;
  }

  let parsed;
  try {
    parsed = parseJsonContent(content);
    const noBetter = validateSuggestionShape(parsed, requiresAlternative);
    if (noBetter) return noBetter;
  } catch {
    const correctionPrompt = [
      'The previous response was invalid JSON, missed required keys, or incorrectly returned noBetterModelFound for a high-risk scenario.',
      'Return only corrected JSON matching the requested schema.',
      requiresAlternative
        ? 'Predictive evidence requires a concrete suggestedScenario. Do not return noBetterModelFound.'
        : 'Return noBetterModelFound only if the predictive evidence is moderate or low risk.',
      `Invalid response: ${content}`,
      `Original request: ${prompt}`
    ].join('\n');
    const retryContent = await callConfiguredAi(correctionPrompt, 260, 0);
    try {
      parsed = parseJsonContent(retryContent);
      const noBetter = validateSuggestionShape(parsed, requiresAlternative);
      if (noBetter) return noBetter;
    } catch {
      throw createAiError(
        'INVALID_AI_RESPONSE',
        'AI suggestion response could not be parsed as valid structured JSON.',
        'The GenAI response did not match the required better-model suggestion schema after one correction retry.',
        502
      );
    }
  }

  return {
    success: true,
    source: env.openSourceAi.provider,
    promptVersion: suggestionPromptVersion,
    model: env.openSourceAi.model,
    ...parsed,
    generatedAt: new Date().toISOString()
  };
}

File: app/backend/src/main/java/com/bms/service/GenAiService.java
package com.bms.service;

import com.bms.config.BmsProperties;
import com.bms.util.ApiException;
import com.bms.util.Json;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestClient;

import java.time.Instant;
import java.util.List;
import java.util.Map;

@Service
public class GenAiService {
  public static final String PROMPT_VERSION = "business-model-rec-v1";
  public static final String SUGGESTION_PROMPT_VERSION = "business-model-suggestion-v1";
  private final BmsProperties properties;
  private final RestClient restClient;
  private final ObjectMapper mapper;

  public GenAiService(BmsProperties properties, RestClient.Builder builder, ObjectMapper mapper) {
    this.properties = properties;
    this.restClient = builder.build();
    this.mapper = mapper;
  }

  public Map<String, Object> generateRecommendation(Map<String, Object> input) {
    if (properties.getAi().getBaseUrl() == null || properties.getAi().getBaseUrl().isBlank() || properties.getAi().getModel() == null || properties.getAi().getModel().isBlank()) {
      return unavailableRecommendation("Configure SAP AI Core or an OpenAI-compatible GenAI endpoint before requesting recommendations.");
    }
    try {
      String content = callAi(buildPrompt(input), 220);
      Map<String, Object> parsed = parseJson(content);
      parsed.put("source", properties.getAi().getProvider());
      parsed.put("promptVersion", PROMPT_VERSION);
      parsed.put("model", properties.getAi().getModel());
      parsed.put("generatedAt", Instant.now().toString());
      return parsed;
    } catch (Exception error) {
      return unavailableRecommendation(error.getMessage());
    }
  }

  public Map<String, Object> generateBetterModelSuggestion(Map<String, Object> input) {
    if (properties.getAi().getBaseUrl() == null || properties.getAi().getBaseUrl().isBlank()) {
      throw new ApiException(HttpStatus.SERVICE_UNAVAILABLE, "GENAI_UNAVAILABLE", "AI suggestion service is not configured.", "Configure SAP AI Core or an OpenAI-compatible GenAI endpoint before requesting suggestions.");
    }
    try {
      Map<String, Object> parsed = parseJson(callAi(buildSuggestionPrompt(input), 1200));
      if (Boolean.TRUE.equals(parsed.get("noBetterModelFound"))) {
        parsed.putIfAbsent("message", "The current model appears optimal based on the available impact and prediction data.");
      } else if (!parsed.containsKey("suggestedScenario") || !parsed.containsKey("expectedBenefits")) {
        throw new ApiException(HttpStatus.BAD_GATEWAY, "INVALID_AI_RESPONSE", "AI suggestion response could not be parsed as valid structured JSON.", "The GenAI response did not match the required better-model suggestion schema after one correction retry.");
      }
      parsed.put("success", true);
      parsed.put("source", properties.getAi().getProvider());
      parsed.put("promptVersion", SUGGESTION_PROMPT_VERSION);
      parsed.put("model", properties.getAi().getModel());
      parsed.put("generatedAt", Instant.now().toString());
      return parsed;
    } catch (ApiException error) {
      throw error;
    } catch (Exception error) {
      throw new ApiException(HttpStatus.SERVICE_UNAVAILABLE, "GENAI_NETWORK_ERROR", "AI suggestion request failed due to a network or provider connection issue.", "Check that the configured GenAI endpoint is reachable and retry.");
    }
  }

  private String callAi(String prompt, int maxTokens) throws Exception {
    Map<String, Object> body = Json.obj("model", properties.getAi().getModel(), "messages", List.of(Json.obj("role", "user", "content", prompt)), "stream", false, "think", false, "format", "json", "options", Json.obj("num_predict", maxTokens, "temperature", 0.2), "max_tokens", maxTokens, "temperature", 0.2, "response_format", Json.obj("type", "json_object"));
    @SuppressWarnings("unchecked")
    Map<String, Object> response = restClient.post().uri(aiEndpointUrl()).header("Content-Type", "application/json").headers(headers -> {
      if (properties.getAi().getApiKey() != null && !properties.getAi().getApiKey().isBlank()) headers.setBearerAuth(properties.getAi().getApiKey());
    }).body(body).retrieve().body(Map.class);
    return parseModelContent(response);
  }

  private String aiEndpointUrl() {
    String url = properties.getAi().getBaseUrl();
    if ("ollama".equals(properties.getAi().getProvider()) && (url.contains("localhost:11434") || url.contains("127.0.0.1:11434"))) {
      return url.replaceAll("/v1/chat/completions/?$", "/api/chat").replaceAll("/v1/?$", "/api/chat");
    }
    return url;
  }

  private String parseModelContent(Map<String, Object> body) throws Exception {
    Object choices = body == null ? null : body.get("choices");
    if (choices instanceof List<?> list && !list.isEmpty()) return Json.str(Json.map(Json.map(list.get(0)).get("message")).get("content"));
    if (body != null && Json.map(body.get("message")).get("content") != null) return Json.str(Json.map(body.get("message")).get("content"));
    if (body != null && body.get("response") != null) return Json.str(body.get("response"));
    if (body != null && body.get("output_text") != null) return Json.str(body.get("output_text"));
    return mapper.writeValueAsString(body);
  }

  @SuppressWarnings("unchecked")
  private Map<String, Object> parseJson(String content) throws Exception {
    String trimmed = content == null ? "" : content.trim();
    trimmed = trimmed.replaceFirst("^```(?:json)?\\s*", "").replaceFirst("\\s*```$", "");
    try {
      return mapper.readValue(trimmed, Map.class);
    } catch (Exception ignored) {
      return Json.obj("executiveSummary", content);
    }
  }

  private String buildPrompt(Map<String, Object> input) {
    return String.join("\n",
        "You are an enterprise commercial model transformation advisor for SAP Order-to-Cash programs.",
        "Return only valid compact JSON. No markdown.",
        "Generate a concise JSON object with keys executiveSummary, recommendation, rationale, risks, assumptions, businessImpact, implementationConsiderations, whyThisModel.",
        "Base every statement on supplied scenario, impact analysis, predictive model output and comparison scores.",
        "Scenario: " + json(input.get("scenario")),
        "Impact analysis: " + json(input.get("impact")),
        "Predictive output: " + json(input.get("prediction")),
        "Predictive historical analysis: " + json(input.get("predictiveAnalysis")),
        "Comparison: " + json(input.get("comparison")));
  }

  private String buildSuggestionPrompt(Map<String, Object> input) {
    return String.join("\n",
        "You are an enterprise SAP Order-to-Cash commercial model architect.",
        "Return only valid JSON. No markdown, no prose outside JSON.",
        "The JSON must include suggestedScenario, improvementSummary, aiRationale, expectedBenefits, o2cImpactChanges, tradeOffs, assumptions and confidence, or noBetterModelFound.",
        "Allowed businessModelType values: subscription, consumption-based, tiered, bundled-services, outcome-based, flexible-funding, hybrid.",
        "Allowed pricingType values: flat, tiered, usage, outcome, bundle, hybrid.",
        "Allowed billingFrequency values: monthly, quarterly, annual, milestone, usage-event.",
        "Original scenario: " + json(input.get("scenario")),
        "Impact analysis: " + json(input.get("impact")),
        "Predictive AI output: " + json(input.get("prediction")),
        "Impact details: " + json(input.get("impactDetails")),
        "Comparison context: " + json(input.get("comparisonContext")));
  }

  private String json(Object value) {
    try { return mapper.writeValueAsString(value); } catch (Exception ignored) { return "{}"; }
  }

  private Map<String, Object> unavailableRecommendation(String details) {
    return Json.obj("success", false, "status", "GENAI_UNAVAILABLE", "source", "unavailable", "promptVersion", PROMPT_VERSION, "model", properties.getAi().getModel(), "message", "AI recommendation is temporarily unavailable.", "customerMessage", "We could not generate a new AI recommendation right now. Predictive impact scores and historical analysis are still available. Please retry after checking the AI service configuration.", "technicalDetails", details, "predictiveOnly", true, "suggestedScenario", null, "executiveSummary", "", "recommendation", "", "rationale", "", "risks", List.of(), "assumptions", List.of(), "businessImpact", "", "implementationConsiderations", "", "whyThisModel", "", "generatedAt", Instant.now().toString());
  }
}
